File: Python/max-number-of-k-sum-pairs.py
# Time:  O(n)
# Space: O(n)

import collections


class Solution(object):
    def maxOperations(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        lookup = collections.Counter()
        for num in nums:
            lookup[num] += 1
        result = 0
        for num, cnt in lookup.iteritems():
            if k-num == num:
                result += cnt//2
            elif k-num in lookup:
                cnt = min(cnt, lookup[k-num])
                result += cnt
                lookup[num] -= cnt
                lookup[k-num] -= cnt
        return result

File: C++/max-number-of-k-sum-pairs.cpp
// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int, int> lookup;
        int result = 0;
        for (const auto& num : nums) {
            if (lookup.count(k - num) && lookup[k - num]) {
                --lookup[k - num];
                ++result;
            } else {
                ++lookup[num];
            }
        }
        return result;
    }
};
